sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc

```diff
@@ -0,0 +1,553 @@
+  template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+  using value_t = property_value<work_group_progress_key, std::integral_constant<forward_progress_guarantee, Guarantee>, std::integral_constant<execution_scope, CoordinationScope>>;
+};
+
+struct sub_group_progress_key {
+  template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+  using value_t = property_value<sub_group_progress_key, std::integral_constant<forward_progress_guarantee, Guarantee>, std::integral_constant<execution_scope, CoordinationScope>>;
+};
+
+struct work_item_progress_key {
+  template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+  using value_t = property_value<work_item_progress_key, std::integral_constant<forward_progress_guarantee, Guarantee>, std::integral_constant<execution_scope, CoordinationScope>>;
+};
+
+template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+inline constexpr work_group_progress_key::value_t<Guarantee, CoordinationScope> work_group_progress;
+
+template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+inline constexpr sub_group_progress_key::value_t<Guarantee, CoordinationScope> sub_group_progress;
+
+template <forward_progress_guarantee Guarantee, execution_scope CoordinationScope>
+inline constexpr work_item_progress_key::value_t<Guarantee, CoordinationScope> work_item_progress;
+
+template <> struct is_property_key<work_group_progress_key> : std::true_type {};
+template <> struct is_property_key<sub_group_progress_key> : std::true_type {};
+template <> struct is_property_key<work_item_progress_key> : std::true_type {};
+
+}
+----
+
+|===
+|Property|Description
+
+|`work_group_progress`
+|The `work_group_progress` property adds the requirement that the kernel must be
+ launched with the specified forward progress guarantees for threads of execution
+ created at `execution_scope::work_group` scope, with the specified
+ coordination scope.
+
+|`sub_group_progress`
+|The `sub_group_progress` property adds the requirement that the kernel must be
+ launched with the specified forward progress guarantees for threads of execution
+ created at `execution_scope::sub_group` scope, with the specified coordination
+ scope.
+
+|`work_item_progress`
+|The `work_item_progress` property adds the requirement that the kernel must be
+ launched with the specified forward progress guarantees for threads of execution
+ created at `execution_scope::work_item` scope, with the specified coordination
+ scope.
+
+|===
+
+
+=== Launch queries [[launch]]
+
+An implementation's ability to satisify requests for specific forward progress
+guarantees for a kernel may be dependent upon: the kernel itself, and the
+features that it uses; the queue (and underlying device) to which the kernel is
+submitted; and the kernel's launch configuration.
+
+It is a user's responsibility to ensure that a kernel requesting specific
```

Comment:
Is the idea for these queries that a user can request an nd-range size, work-group size, and sub-group size in order to satisfy the forward progress guarantees requested for that kernel?

Could there be situations where a user places strong enough forward progress guarantee requirements that there is no valid value that can be returned from a query? Perhaps the queries could just return `1` for all of them.

But then I could imagine a case where other kernel properties could conflict with this, for example, if a kernel is defined with a forward progress guarantee, but also a required work-group size. I'm not sure if we could provide a compile-time error for this, so we'd need some way to reflect this in the result of these queries, perhaps a query for whether the kernel is executable on the device at all with the requested forward progress guarantees.

Reply:
Right, exactly.  Returning `1` for all the queries is exactly the fallback behavior I had in mind.

I hadn't considered the case where the user might specify an invalid combination of properties.  I _think_ in the specific case you mentioned you'd probably already get a runtime error as a result of trying to launch the kernel with a different number of work-groups than the required work-group size.

As a general solution, I think we could just say that `max_num_work_groups` is `0` if the kernel cannot be launched at all.  I like this because it doesn't require a new query, and because it would also provide us a way to detect if the user specified an amount of dynamic memory that is incompatible with the kernel.  For example, even without any forward progress guarantees, something like `kernel.get_info<max_num_work_groups>(q, range{1, 1, 1}, MAX_INT)` could return `0`.

Reply:
Ben also suggested below that the behavior of the launch queries was unclear.  While addressing that comment in da6d477d9257f69b8e1d047cbeaf1a411a9a6462, I added some draft language for the `0`/`1` solution described above.